### PR TITLE
rely upon gas fee controller for gas price estimates

### PR DESCRIPTION
**Pull request by brad-decker**
Changes
* returns `false` from getNetworkEIP1559Compatibility if the `SHOW_EIP_1559_UI` environment variable isn't true. This allows testnets such as Ropsten, Goerli, Rinkeby, etc that have launched EIP-1559 to still get good `gasPrice` estimates until we release our UI elements.
* Plugs in values from the `GasFeeController` state into the custom-gas selectors. These selectors will be removed in favor of directly accessing the underlying state values (or creating new, more apt named selectors) in a future PR.
* updates tests to have mock data appropriate for these changes. I attempted to only modify test mock data or remove tests that were no longer relevant (covering removed code)
* Adds calls to start polling on the gasFeeController from a number of places that were previously calling the fetch method from gas.duck.js.
* Removes all fetch related code from gas.duck.js, it now ONLY exists to store custom data which will also go away as we get closer to merging our EIP-1559 work. 

Items remaining
- [x] The gasPrice on send page is being constantly overwritten with values. I need to tap into the customGasPrice state to see if it has been set and then avoid changing gasPrice once that happens. We'll need to figure this out for EIP-1559 too because the maxFeePerGas/gasPrice will update periodically but **isn't visible to the user**, resulting in possibly randomly displaying insufficient funds warnings? 
- [x] Failing test that I'm not sure why is happening. 

### QA / Testing Steps
- Use the send flow on various networks and ensure a reasonable gasPrice is set. this is 1 on rinkeby, 5 on Goerli, 10 on Kovan, 2 on Ropsten (as of this writing 7/14/2021 1:20 CST).
- On the send flow, on custom networks, switch networks and see that the gasPrice updates within a few seconds (no more than 10)
- On the send flow, if you change your gasPrice, your price does not change every 10 seconds.
- Confirmation edit gas price logic works on mainnet (I *believe* that custom networks do not use estimates they set the gasPrice to the transaction gasPrice, the only place estimates are displayed is in the buttons)
- speed up/cancel works

### Known Issues
<details>
<summary>Gas customization modal on mainnet does not update gasPrice values when estimates change</summary>

https://user-images.githubusercontent.com/4448075/125673410-f903853e-933d-4f5b-afb9-5cb81fe72ae7.mov

<br/>
<ul>
<li>I don't know how much time we should spend resolving this as it doesn't actually change the *result* of the behavior from mainnet. If the user had opened this window the gasPrices wouldn't have updated but it doesn't mean the expectations of the network hadn't changed.</li>
<li>Fixing this will require introducing a whole layer of complexity to the existing gas modal that we intend on deprecating soon</li>
</ul>

</details>

### CODE REVIEW

File: app/scripts/metamask-controller.js
Comment (danjm) [inline app/scripts/metamask-controller.js:2061]: console.log

File: app/scripts/controllers/network/network.js
Comment (danjm) [inline app/scripts/controllers/network/network.js:162]: :+1:

File: ui/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js
Comment (danjm) [inline ui/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.container.js:10]: :+1:

File: ui/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.component.js
Comment (danjm) [inline ui/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.component.js:56]: Not sure if there is anything we can do about this:

Is it possible for the component to unmount before `getGasFeeEstimatesAndStartPolling()` resolves, thereby polling having been initiated in the controller but disconnect won't be called here, leaving the controller polling indefinitely?

File: ui/helpers/utils/conversions.util.js
Comment (danjm) [inline ui/helpers/utils/conversions.util.js:83]: It seems that this new `numberOfDecimals` parameter is not used anywhere

Comment (brad-decker) [reply ui/helpers/utils/conversions.util.js:83]: Yes, needs to be removed didn't end up needing it!

File: ui/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.component.js
Comment (brad-decker) [reply ui/components/app/gas-customization/gas-modal-page-container/gas-modal-page-container.component.js:56]: Hmmm.... good point. I think that affects our `useGasEstimates` hook, too. At least for this component, it won't matter long term because I plan on ceremoniously deleting it once @darkwing's efforts are fully realized. :D 

For the useGasEstimates hook we can do something like this 

```
useEffect(() => {
  let pollingToken;
  let isMounted = true;
  getGasFeeEstimatesAndStartPolling().then((token) => {
    if (!isMounted) {
      return disconnectMethodNameThatIForgot(token);
    }
    pollingToken = token;
  });
  return () => {
    isMounted = false;
    if (pollingToken) {
     //disconnect
    }
  }
}, []);
```

File: ui/pages/confirm-transaction/confirm-transaction.container.js
Comment (danjm) [inline ui/pages/confirm-transaction/confirm-transaction.container.js:6]: :+1:

File: ui/pages/confirm-send-ether/confirm-send-ether.container.js
Comment (danjm) [inline ui/pages/confirm-send-ether/confirm-send-ether.container.js:17]: :+1:

File: ui/pages/confirm-send-ether/confirm-send-ether.component.js
Comment (danjm) [inline ui/pages/confirm-send-ether/confirm-send-ether.component.js:16]: :+1:

File: ui/hooks/useRetryTransaction.js
Comment (danjm) [inline ui/hooks/useRetryTransaction.js:31]: This might be a bit nitpicky: it feels like it would be nice to separate the "get gas fee estimates" concern from the "start polling" concern. Calling `useGasFeeEstimates()` in some cases for the returned data, and in other cases for the side effect of starting polling, seems like it could be a bit confusing to future readers/maintainers

Comment (danjm) [reply ui/hooks/useRetryTransaction.js:31]: But also, why is `useGasFeeEstimates` called here and not in a component?

Comment (danjm) [reply ui/hooks/useRetryTransaction.js:31]: And is this also needed for cancels?

Comment (brad-decker) [reply ui/hooks/useRetryTransaction.js:31]: agreed, but the good news is this is a temporary measure until users are using the new UI.

Comment (brad-decker) [reply ui/hooks/useRetryTransaction.js:31]: I can just copy/paste the polling code from that hook here instead

File: ui/ducks/gas/gas.duck.js
Comment (danjm) [inline ui/ducks/gas/gas.duck.js:197]: This had a guard against trying to retry transaction with a `0x0` price. I guess that is unnecessary now because the price will always come from the controller?

File: app/scripts/metamask-controller.js
Comment (danjm) [inline app/scripts/metamask-controller.js:194]: :+1:

File: ui/pages/confirm-transaction/confirm-transaction.component.js
Comment (danjm) [inline ui/pages/confirm-transaction/confirm-transaction.component.js:74]: The same question I raised elsewhere about cancelling polling can be raised here.

File: ui/hooks/useRetryTransaction.js
Comment (brad-decker) [reply ui/hooks/useRetryTransaction.js:31]: > But also, why is useGasFeeEstimates called here and not in a component?

I'm having a hard time remembering why I did this, though it is fine to call a hook from within a hook. Thats part of their power in composability. I do think here it'd be more appropriate to just poll -- however, like I said I'm having hard time remembering why this was necessary.

I did remove the setCustomGasPriceForRetry method that calls the fetch code, but I think this is covered by having a polling creator in the gas modal so this likely can just be removed.

File: ui/ducks/gas/gas.duck.js
Comment (brad-decker) [reply ui/ducks/gas/gas.duck.js:197]: yeah, that's correct.

File: ui/ducks/send/send.js
Comment (danjm) [inline ui/ducks/send/send.js:192]: Before this would be `options.blockGasLimit` if it was defined. Now it will be `GAS_LIMITS.BASE_TOKEN_ESTIMATE` if `options.blockGasLimit` is defined. To ensure backwards compatibility while addressing the issue this is meant to address, I might suggest:
```
let blockGasLimit = MIN_GAS_LIMIT_HEX;
if (options.blockGasLimit) {
blockGasLimit = options.blockGasLimit;
} else if (sendToken) {
blockGasLimit = GAS_LIMITS.BASE_TOKEN_ESTIMATE;
}
```

File: ui/ducks/send/send.js
Comment (danjm) [inline ui/ducks/send/send.js:345]: We should investigate the difference between `blockGasLimit` and `currentBlockGasLimit` and why we were using something that was seemingly not defined. We don't necessarily have to do this before this PR is merged, but I would like to create an issue, and block release on answering the question, to make sure there are no other related bugs, or no important differences between what we expected `blockGasLimit` to be and what `currentBlockGasLimit` is

File: ui/ducks/send/send.js
Comment (danjm) [inline ui/ducks/send/send.js:378]: Nit: feels like this should be be in some `util` type file

**Review by danjm — CHANGES_REQUESTED** [review]
Completed review. Overall everything makes sense and looks good. Once my comments and questions are addressed we are good.

File: ui/ducks/send/send.js
Comment (brad-decker) [reply ui/ducks/send/send.js:378]: I disagree, as it will only ever be used here I hope, and maybe will eventually be removed entirely. Moving things to Util files is a vote of confidence for their usage and it is not exported on purpose.

File: ui/ducks/send/send.js
Comment (brad-decker) [reply ui/ducks/send/send.js:345]: It most likely happened as a result of the send flow refactor where we were taking currentBlockGasLimit from state and passing it to utility methods as a 'blockGasLimit' prop and then I improperly selected it from state.

Comment (brad-decker) [reply ui/ducks/send/send.js:345]: currentBlockGasLimit is the legitimate value. A prop named blockGasLimit is used in the tx-gas-utils and transaction controller, but its set in state in the `account-tracker` as `currentBlockGasLimit`

```
  async _updateForBlock(blockNumber) {
    this._currentBlockNumber = blockNumber;

    // block gasLimit polling shouldn't be in account-tracker shouldn't be here...
    const currentBlock = await this._query.getBlockByNumber(blockNumber, false);
    if (!currentBlock) {
      return;
    }
    const currentBlockGasLimit = currentBlock.gasLimit;
    this.store.updateState({ currentBlockGasLimit });

    try {
      await this._updateAccounts();
    } catch (err) {
      log.error(err);
    }
  }
```

File: ui/ducks/send/send.js
Comment (brad-decker) [reply ui/ducks/send/send.js:192]: if options.blockGasLimit is defined it'll be set to that due to the logical `||` operator. the asset type is only checking if the blockGasLimit is not provided

Comment (brad-decker) [reply ui/ducks/send/send.js:192]: I can make this change for legibility though.

Comment (brad-decker) [reply ui/ducks/send/send.js]: Done in latest

File: ui/pages/confirm-transaction/confirm-transaction.component.js
Comment (brad-decker) [reply ui/pages/confirm-transaction/confirm-transaction.component.js:75]: I added a instance variable _isMounted where this occurs, and if _isMounted if alse after the promise resolves we disconnect instead of setting the pollingToken in state.

File: ui/hooks/useRetryTransaction.js
Comment (brad-decker) [reply ui/hooks/useRetryTransaction.js]: removed in latest

**Comment by brad-decker** [comment]
I think I resolved all of the feedback @danjm

**Review by danjm — APPROVED** [review]
LGTM

**Comment by PeterYinusa** [comment]
LGTM!

**Review by darkwing — APPROVED** [review]

**Pull request merged**
